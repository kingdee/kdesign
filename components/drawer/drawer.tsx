import React, {
  ReactNode,
  useContext,
  useEffect,
  useState,
  ReactElement,
  useCallback,
  CSSProperties,
  createRef,
  useRef,
} from 'react'
import { Icon } from '../index'
import ReactDOM from 'react-dom'
import { tuple } from '../_utils/type'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import devWarning from '../_utils/devwarning'
import { useResizeObserver, useOverflowHidden } from '../_utils/hooks'
// import { usePrevious } from '../_utils/hooks'
import { toArray } from '../_utils/react-children'
export const PlacementTypes = tuple('top', 'right', 'bottom', 'left')
export type PlacementType = typeof PlacementTypes[number]
const relativePostionTypes = ['relative', 'absolute', 'fixed']
type CSSSelector = string
const horizontalPlacements = ['left', 'right']
const verticalPlacements = ['top', 'bottom']

/**
 * TODO:
 * 多层抽屉的实现方式无法确定（需求不明确，没有现行的业务场景）: 思考的两种方式:
 * 1. context嵌套，采用ant-design的context嵌套来实现，需要一个ref存父级Drawer（存疑, 但应该也可以）
 * 2. 状态存储，全局Map记录，如果不想记录可以用props来绕过，需要ref记录当前显示的次序（数组下标）
 */
// type placementSizeMap = {
//   height?: number
//   width?: number
// }
// interface PlacementMap<T> {
//   [prop: string]: T[]
// }
// type DrawerMapTypeByPlacementAndContainer = WeakMap<HTMLElement, PlacementMap<placementSizeMap>>
// const drawersMap: DrawerMapTypeByPlacementAndContainer = new WeakMap()

export interface IDrawerProps {
  className?: string
  closable?: boolean
  closeIcon?: ReactNode
  children?: ReactNode
  destroyOnClose?: boolean
  footer?: ReactNode
  footerClassName?: string
  footerStyle?: CSSProperties
  forceRender?: boolean
  getContainer?: (() => Element | CSSSelector | false | null) | Element | false
  headerStyle?: CSSProperties
  headerClassName?: string
  keyboard?: boolean
  mask?: boolean
  maskClassName?: string
  maskClosable?: boolean
  maskStyle?: CSSProperties
  onClose?: () => void
  placement?: PlacementType
  prefixCls?: string
  title?: ReactNode
  titleClassName?: string
  titleStyle?: CSSProperties
  visible?: boolean
  width?: string | number
  height?: string | number
  zIndex?: string | number
  disableScroll?: boolean
}
const documentBody = document.body

// function transformSize(size: string | number, placement: string) {
//   if (typeof size === 'string') {
//     // 只考虑px 或者 %
//     devWarning(
//       !(size.endsWith('px') && size.endsWith('%')),
//       'drawer',
//       `cannot calculate drawer size: width | height for '${size}' besides px or %`,
//     )
//     if (placement === 'right' || placement === 'bottom') return size || 0
//     return size ? '-' + size : 0
//   }
//   if (placement === 'right' || placement === 'bottom') return (size || 0) + 'px'
//   return ('-' + size || 0) + 'px'
// }

const InternalDrawer = (props: IDrawerProps, ref: unknown): ReactElement | React.ReactPortal => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const drawerProps: IDrawerProps = getCompProps('Drawer', userDefaultProps, props) // 属性需要合并一遍用户定义的默认属性
  const {
    className,
    closable,
    closeIcon,
    children,
    onClose,
    keyboard,
    mask,
    maskClosable,
    maskStyle,
    maskClassName,
    placement,
    title,
    titleStyle,
    titleClassName,
    footer,
    footerStyle,
    footerClassName,
    visible,
    width,
    height,
    headerStyle,
    headerClassName,
    zIndex: zindex,
    getContainer,
    forceRender,
    destroyOnClose,
    prefixCls: customPrefixcls,
    disableScroll,
  } = drawerProps
  const drawerPrefixCls = getPrefixCls!(prefixCls, 'drawer', customPrefixcls)
  const [showChildren, setShowChildren] = useState(false)
  const [currentWidth, setCurrentWidth] = useState(width)
  // const previousWidthRef = usePrevious(currentWidth)
  // const previousHeightRef = usePrevious(currentHeight)
  const containerRef = (ref as any) || createRef<HTMLDivElement>()
  const contentRef = useRef<HTMLDivElement>(null)

  devWarning(PlacementTypes.indexOf(placement!) === -1, 'drawer', `cannot found drawer type '${placement}'`)

  let drawerContainer: Element | CSSSelector | undefined | null | false =
    typeof getContainer === 'function' ? getContainer() : getContainer
  const isAtOriginalPlace = drawerContainer === false
  if (!(drawerContainer instanceof Element) && !isAtOriginalPlace) {
    try {
      drawerContainer = document.querySelector(drawerContainer as CSSSelector)
    } catch (error) {
      devWarning(true, 'drawer', 'getContainer"s returnValue/value is not a correct CSSSelector')
    }
    drawerContainer = documentBody
  }
  const isBody = (!drawerContainer && !isAtOriginalPlace) || drawerContainer === document.body
  useEffect(() => {
    setShowChildren(!!forceRender)
  }, [forceRender])
  useEffect(() => {
    let node: HTMLElement | null = null
    const portalNode = containerRef.current
    // 应该只有可见才能影响
    // 但是这样会有闪烁的问题
    if (!isBody && drawerContainer) {
      node = drawerContainer as HTMLElement
    } else if (isAtOriginalPlace) {
      node = portalNode && portalNode.parentElement
    }
    if (node && !relativePostionTypes.includes(window.getComputedStyle(node).position)) {
      devWarning(true, 'drawer', 'getContainer"s returnValue/value is not a relative/absolute/fixed positioned DOM')
    }
  }, [drawerContainer, isBody, isAtOriginalPlace, containerRef])

  const handleContainerResize = useCallback(
    ({ width: containerWidth }) => {
      if (!visible && width === currentWidth) return
      if (containerWidth < currentWidth!) {
        // 这里改成操作DOM？
        setCurrentWidth(containerWidth)
        // 这里拿不到正确的之前的宽度
        //  if (width !== previousWidthRef!)
        // 会一直触发
      } else {
        setCurrentWidth(width)
      }
    },
    [width, visible, currentWidth],
  )
  useResizeObserver(() => (drawerContainer || containerRef.current?.parentNode) as HTMLElement, handleContainerResize)
  useOverflowHidden(
    (drawerContainer || containerRef.current?.parentNode) as HTMLElement,
    visible && mask,
    disableScroll,
  )

  // 为了单测
  useEffect(() => {
    setCurrentWidth(width)
  }, [width])

  const closeDrawer = useCallback(() => {
    onClose && onClose()
    destroyOnClose && setShowChildren(false)
  }, [onClose, destroyOnClose])

  const escapeToCloseModal = useCallback(
    (evt: KeyboardEvent) => {
      // In Internet Explorer 9 and Firefox 36 and earlier, the Esc key returns "Esc" instead of "Escape"
      if (['Escape', 'Esc'].includes(evt.key)) {
        closeDrawer()
      }
    },
    [closeDrawer],
  )

  useEffect(() => {
    if (keyboard) {
      document.body.addEventListener('keydown', escapeToCloseModal)
    }
    return () => {
      document.body.removeEventListener('keydown', escapeToCloseModal)
    }
  }, [keyboard, closeDrawer, escapeToCloseModal])

  const handleMaskClick = useCallback(() => {
    if (maskClosable) {
      closeDrawer()
    }
  }, [maskClosable, closeDrawer])

  const getHorizontalBoolAndPlacementName = useCallback(() => {
    const isHorizontal = placement === 'left' || placement === 'right'
    const placementName = `translate${isHorizontal ? 'X' : 'Y'}`
    return {
      isHorizontal,
      placementName,
    }
  }, [placement])

  const setDraerTransform = useCallback(
    (_isOpen?: boolean, _placementName?: string, _value?: string | number) => {
      if (!contentRef.current) return
      contentRef.current.style.transform = ''
    },
    [contentRef],
  )

  useEffect(() => {
    const { isHorizontal, placementName } = getHorizontalBoolAndPlacementName()
    const contentValue = contentRef.current
      ? contentRef.current.getBoundingClientRect()[isHorizontal ? 'width' : 'height']
      : 0
    const value = (isHorizontal ? width : height) || contentValue
    if (visible) {
      setDraerTransform(visible, placementName, value)
    }
  }, [contentRef, getHorizontalBoolAndPlacementName, height, setDraerTransform, visible, width])

  const drawerClasses = classNames(drawerPrefixCls, className, {
    [`${drawerPrefixCls}-hide`]: !visible,
    [`${drawerPrefixCls}-active`]: visible,
    [`${drawerPrefixCls}-not-at-body`]: !isBody,
  })
  const headerClass = `${drawerPrefixCls}-header`

  const { placementName } = getHorizontalBoolAndPlacementName()
  const placementPos = placement === 'left' || placement === 'top' ? '-100%' : '100%'
  const transform = visible ? '' : `${placementName}(${placementPos})`
  const containerStyle = Object.assign(
    { [placement!]: 0 },
    horizontalPlacements.includes(placement!)
      ? {
          width: width,
          transform: transform,
        }
      : {
          height,
          transform: transform,
        },
  )
  const container = (
    <div
      className={classNames({
        [`${drawerPrefixCls}-container-box`]: true,
        [`${drawerPrefixCls}-has-container-box`]: drawerContainer,
        [`${drawerPrefixCls}-container-${placement}`]: true,
      })}
      style={containerStyle}
      ref={contentRef}
    >
      {(title || closable) && (
        <div className={classNames(headerClass, headerClassName)} style={headerStyle}>
          <div className={classNames(`${drawerPrefixCls}-title`, titleClassName)} style={titleStyle}>
            {title}
          </div>
          {closable && (
            <div className={`${drawerPrefixCls}-close-icon`} onClick={() => closeDrawer()}>
              {closeIcon || <Icon type="close" />}
            </div>
          )}
        </div>
      )}
      <div className={`${drawerPrefixCls}-body`}>{(showChildren || visible) && toArray(children)}</div>
      {footer && (
        <div className={classNames(`${drawerPrefixCls}-footer`, footerClassName)} style={footerStyle}>
          {footer}
        </div>
      )}
    </div>
  )
  const opposition: string =
    !mask && PlacementTypes.includes(placement as PlacementType)
      ? horizontalPlacements.includes(placement as PlacementType)
        ? horizontalPlacements.filter((p) => p !== placement)[0]
        : verticalPlacements.filter((p) => p !== placement)[0]
      : ''
  const noneMaskStyle = !mask
    ? {
        width: containerStyle.width,
      }
    : {}
  const comp: ReactElement = (
    <div
      className={drawerClasses}
      style={{
        zIndex: (zindex as any) || 'auto',
        ...noneMaskStyle,
        ...(opposition ? { [opposition]: 'unset' } : {}),
      }}
      ref={containerRef}
    >
      {/* 增加clickOutside */}
      {mask && (
        <div
          onClick={handleMaskClick}
          className={classNames(`${drawerPrefixCls}-mask`, maskClassName)}
          style={maskStyle}
        ></div>
      )}
      {container}
    </div>
  )

  if (drawerContainer instanceof HTMLElement) return ReactDOM.createPortal(comp, drawerContainer!)
  else if (drawerContainer !== false) {
    return ReactDOM.createPortal(comp, documentBody!)
  }
  return comp
}
const Drawer = React.forwardRef<unknown, IDrawerProps>(InternalDrawer)

Drawer.displayName = 'Drawer'
export default Drawer
