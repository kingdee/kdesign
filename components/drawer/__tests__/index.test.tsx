import React from 'react'
import { mount, render, ReactWrapper } from 'enzyme'
import Drawer from '../index'
import { PlacementType, PlacementTypes } from '../drawer'
// import { act } from 'react-dom/test-utils'
import mountTest from '../../../tests/shared/mountTest'
import { simulateEvent } from '../../../tests/shared/simulateEvent'
import ConfigProvider from '../../config-provider/index'

jest.useFakeTimers()
const setWrapperVisibleFC = () => {
  return (wrapper: ReactWrapper) => {
    wrapper.setProps({
      visible: true,
    })
    wrapper.update()
  }
}
const prefixClassName = 'kd-drawer'
const classNameMap = {
  closeIcon: `.${prefixClassName}-close-icon`,
  mask: `.${prefixClassName}-mask`,
  title: `.${prefixClassName}-title`,
  containerBox: `.${prefixClassName}-container-box`,
  body: `.${prefixClassName}-body`,
  footer: `.${prefixClassName}-footer`,
  box: `.${prefixClassName}`,
  hideBox: `.${prefixClassName}-hide`,
}

const unmountHelp = (wrapper: ReactWrapper) => {
  expect(() => {
    wrapper.unmount()
  }).not.toThrow()
}
describe('Drawer', () => {
  // 1. mount test
  mountTest(Drawer)
  PlacementTypes.forEach((type) => {
    mountTest(() => <Drawer placement={type} />)
  })
  // 2. render test
  it('renders correctly', () => {
    expect(render(<Drawer getContainer={false} />)).toMatchSnapshot()
    PlacementTypes.forEach((type) => {
      expect(render(<Drawer placement={type} getContainer={false} />)).toMatchSnapshot()
    })
  })
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const type = 'who am I' as any as PlacementType
    render(<Drawer placement={type} getContainer={false} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-drawer: cannot found drawer type 'who am I'")
  })
  // 3. render no child without errors
  it('render no child without errors', () => {
    expect(mount(<Drawer></Drawer>)).toMatchSnapshot()
  })
  // 4. render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Drawer>
          {null}
          {undefined}
        </Drawer>,
      ),
    ).toMatchSnapshot()
  })
  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Drawer></Drawer>)
    expect((wrapper.type() as any).displayName).toBe('Drawer')
  })

  const customedReactNodeProp = ['title', 'footer', 'closeIcon']
  const generatorCustomedReactNode = (prop: string) => {
    return [
      prop,
      () => {
        return <div className={`customed-${prop}`}></div>
      },
      `.customed-${prop}`,
    ]
  }
  it.each(customedReactNodeProp.map(generatorCustomedReactNode))(
    'should %s use right',
    (prop: string, CustomedNode, className) => {
      const wrapper = mount(<Drawer {...{ [prop]: <CustomedNode /> }} />)
      expect(wrapper.exists(className)).toBe(true)
      expect(() => {
        wrapper.unmount()
      }).not.toThrow()
    },
  )
  // 6. api
  // className footerClassName headerClassName maskClassName titleClassName
  it('should className use right', () => {
    const wrapper = mount(
      <Drawer
        visible
        maskClassName={'test-mask'}
        className={'test-over'}
        maskStyle={{ color: 'red' }}
        title={'test'}
        titleClassName={'test-title'}
        titleStyle={{ color: 'black' }}
        footer={'footer'}
        footerClassName={'test-footer'}
        footerStyle={{ color: 'blue' }}
      />,
    )
    expect(wrapper.exists('.test-mask')).toBe(true)
    expect(wrapper.exists('.test-title')).toBe(true)
    expect(wrapper.exists('.test-footer')).toBe(true)
    expect(wrapper.exists('.test-over')).toBe(true)
    expect((wrapper.find(classNameMap.mask).getDOMNode() as HTMLElement).style.color).toBe('red')
    expect((wrapper.find(classNameMap.title).getDOMNode() as HTMLElement).style.color).toBe('black')
    expect((wrapper.find(classNameMap.footer).getDOMNode() as HTMLElement).style.color).toBe('blue')

    unmountHelp(wrapper)
  })
  // closable
  it('should closable use right', () => {
    const wrapper = mount(<Drawer closable={false} />)
    expect(wrapper.exists(classNameMap.closeIcon)).toBe(false)
    unmountHelp(wrapper)
  })
  // keyboard
  it('should keyboard use right', () => {
    const onClose = () => {
      wrapper.setProps({
        visible: false,
      })
    }
    const wrapper = mount(<Drawer keyboard={true} visible onClose={onClose} />)
    simulateEvent(document.body, 'keydown', {
      key: 'Escape',
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(true)
    const setWrapperVisible = setWrapperVisibleFC()
    setWrapperVisible(wrapper)
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    simulateEvent(document.body, 'keydown', {
      key: 'Esc',
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(true)
    setWrapperVisible(wrapper)
    simulateEvent(document.body, 'keydown', {
      key: 'enter',
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    wrapper.setProps({
      keyboard: false,
    })
    wrapper.update()
    simulateEvent(document.body, 'keydown', {
      key: 'Escape',
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    simulateEvent(document.body, 'keydown', {
      key: 'Esc',
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    unmountHelp(wrapper)
  })
  // maskClosable
  it('should maskClosable use right', () => {
    const onClose = () => {
      wrapper.setProps({
        visible: false,
      })
    }
    const wrapper = mount(<Drawer maskClosable visible onClose={onClose} />)
    wrapper.find(classNameMap.mask).simulate('click')
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(true)
    const setWrapperVisible = setWrapperVisibleFC()
    setWrapperVisible(wrapper)
    wrapper.setProps({
      maskClosable: false,
    })
    wrapper.find(classNameMap.mask).simulate('click')
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    unmountHelp(wrapper)
  })
  // mask
  it('should mask use right', () => {
    const wrapper = mount(<Drawer mask={false} visible />)
    expect(wrapper.exists(classNameMap.mask)).toBe(false)

    wrapper.setProps({
      mask: true,
    })
    wrapper.update()
    expect(wrapper.exists(classNameMap.mask)).toBe(true)
    unmountHelp(wrapper)
  })
  // visible
  it('should visible use right', () => {
    const onClose = () =>
      setTimeout(() => {
        wrapper.setProps({
          visible: false,
        })
      }, 5000)
    const wrapper = mount(<Drawer visible onClose={onClose} />)
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    wrapper.find(classNameMap.closeIcon).simulate('click')
    wrapper.update()
    jest.advanceTimersByTime(4000)
    expect(wrapper.exists(classNameMap.hideBox)).toBe(false)
    jest.advanceTimersByTime(1000)
    wrapper.update()
    expect(wrapper.exists(classNameMap.hideBox)).toBe(true)
    wrapper.update()

    unmountHelp(wrapper)
  })
  // width
  it.each([['width', ['left', 'right']]])('should %s use right', (prop: 'width', placementTypesList) => {
    PlacementTypes.forEach((type) => {
      const wrapper = mount(<Drawer visible {...{ [prop]: 500 }} placement={type} />)
      const node = () => wrapper.find(classNameMap.containerBox).getDOMNode() as HTMLElement
      if (placementTypesList.includes(type)) {
        expect(node().style[prop]).toBe('500px')
        wrapper.setProps({
          [prop]: '60%',
        })
        expect((wrapper.find(classNameMap.containerBox).getDOMNode() as HTMLElement).style[prop]).toBe('60%')
        unmountHelp(wrapper)
      } else {
        expect(node().style[prop]).not.toBe('500px')
        wrapper.setProps({
          [prop]: '60%',
        })
        expect((wrapper.find(classNameMap.containerBox).getDOMNode() as HTMLElement).style[prop]).not.toBe('60%')
        unmountHelp(wrapper)
      }
    })
  })
  // zIndex
  it('should zIndex use right', () => {
    const wrapper = mount(<Drawer visible width={500} zIndex={10} />)
    expect((wrapper.find(classNameMap.box).getDOMNode() as HTMLElement).style.zIndex).toBe('10')
    unmountHelp(wrapper)
  })

  // get container
  it('should getContainer use right', () => {
    // 通过relative来验证
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const TestComp = () => {
      const ref = React.useRef<HTMLDivElement>(null)
      return (
        <div ref={ref} className="test-ref">
          <Drawer getContainer={() => ref.current} visible />
        </div>
      )
    }
    const renderComp = mount(<TestComp />)
    // 需要setProps一下才触发
    renderComp.setProps({})
    renderComp.update()
    // expect(window.getComputedStyle((renderComp.find('.test-ref').getDOMNode() as HTMLElement)).getPropertyValue('position')).toBe('relative')
    // expect((renderComp.find('.test-ref').getDOMNode() as HTMLElement).style.position).toBe('relative')
    expect(mockWarn.mock.calls[0][0]).toMatch(
      'Warning: [kdesign]-drawer: getContainer"s returnValue/value is not a relative/absolute/fixed positioned DOM',
    )
    const AnotherTestComp = () => {
      const ref = React.useRef(null)
      return (
        <div ref={ref} className="test-body">
          <Drawer getContainer={() => null} visible />
        </div>
      )
    }
    const anotherRenderComp = mount(<AnotherTestComp />)
    anotherRenderComp.setProps({})
    anotherRenderComp.update()
    // expect((anotherRenderComp.find('.test-body').getDOMNode() as HTMLElement).style.position).not.toBe('relative')
    // const ClassTestComp = () => {
    //   return (
    //     <div className={'test-get-container'}>
    //       <Drawer getContainer={() => ((document.querySelector('.test-get-container') as HTMLElement) || '.test-get-container')} visible />
    //     </div>
    //   )
    // }
    // const classRenderComp = mount(<ClassTestComp />)
    // 这个不管用
    // classRenderComp.setProps({
    // })
    // classRenderComp.update()
    // expect((classRenderComp.find('.test-get-container').getDOMNode() as HTMLElement).style.position).toBe('relative')

    const ErrorTestComp = () => {
      return (
        <div className={'test-get-container'}>
          <Drawer getContainer={() => ':::'} visible />
        </div>
      )
    }
    const errorRenderComp = mount(<ErrorTestComp />)
    errorRenderComp.setProps({})
    errorRenderComp.update()
    expect((errorRenderComp.find('.test-get-container').getDOMNode() as HTMLElement).style.position).not.toBe(
      'relative',
    )
    expect(mockWarn.mock.calls[1][0]).toMatch(
      'Warning: [kdesign]-drawer: getContainer"s returnValue/value is not a correct CSSSelector',
    )
  })
  // forceRender
  it('should forceRender use right', () => {
    const onClose = () => {
      wrapper.setProps({
        visible: false,
      })
    }
    const wrapper = mount(
      <Drawer onClose={onClose}>
        <div className="test-force" />
      </Drawer>,
    )
    expect(wrapper.exists('.test-force')).toBe(false)
    wrapper.setProps({
      visible: true,
    })
    expect(wrapper.exists('.test-force')).toBe(true)
    wrapper.find(classNameMap.closeIcon).simulate('click')
    wrapper.update()
    expect(wrapper.exists('.test-force')).toBe(false)
    wrapper.setProps({
      forceRender: true,
    })
    wrapper.update()
    expect(wrapper.exists('.test-force')).toBe(true)
    wrapper.setProps({
      visible: true,
    })
    expect(wrapper.exists('.test-force')).toBe(true)
    wrapper.find(classNameMap.closeIcon).simulate('click')
    wrapper.update()
    expect(wrapper.exists('.test-force')).toBe(true)
    unmountHelp(wrapper)
  })
  // destroyOnClose
  it('should destroyOnClose, use right', () => {
    const onClose = () => {
      wrapper.setProps({
        visible: false,
      })
    }
    const wrapper = mount(
      <Drawer forceRender destroyOnClose visible onClose={onClose}>
        <div className="test-force" />
      </Drawer>,
    )
    expect(wrapper.exists('.test-force')).toBe(true)
    wrapper.find(classNameMap.closeIcon).simulate('click')
    wrapper.update()
    // destroyOnClose优先级高
    expect(wrapper.exists('.test-force')).toBe(false)

    unmountHelp(wrapper)
  })

  // ConfigProvider
  it('should config use config provider', () => {
    const modalConfig = {
      compDefaultProps: {
        Drawer: {
          placement: 'left',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={modalConfig}>
        <Drawer></Drawer>
      </ConfigProvider>,
    )
    expect(wrapper.find(classNameMap.containerBox)).toHaveClassName(`.${prefixClassName}-container-left`)
  })
})
