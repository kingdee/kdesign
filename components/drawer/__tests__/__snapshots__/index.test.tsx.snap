// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer render no child without errors 1`] = `
<Drawer>
  <Portal
    containerInfo={
      <body>
        <div
          class="kd-drawer kd-drawer-hide"
          style="z-index: 1050;"
        >
          <div
            class="kd-drawer-mask"
          />
          <div
            class="kd-drawer-container-box kd-drawer-has-container-box kd-drawer-container-right"
            style="right: 0px; width: 320px; transform: translateX(100%);"
          >
            <div
              class="kd-drawer-header"
            >
              <div
                class="kd-drawer-title"
              />
              <div
                class="kd-drawer-close-icon"
              >
                <i
                  class="kdicon kdicon-close"
                />
              </div>
            </div>
            <div
              class="kd-drawer-body"
            />
          </div>
        </div>
      </body>
    }
  >
    <div
      className="kd-drawer kd-drawer-hide"
      style={
        Object {
          "zIndex": 1050,
        }
      }
    >
      <div
        className="kd-drawer-mask"
        onClick={[Function]}
      />
      <div
        className="kd-drawer-container-box kd-drawer-has-container-box kd-drawer-container-right"
        style={
          Object {
            "right": 0,
            "transform": "translateX(100%)",
            "width": 320,
          }
        }
      >
        <div
          className="kd-drawer-header"
        >
          <div
            className="kd-drawer-title"
          />
          <div
            className="kd-drawer-close-icon"
            onClick={[Function]}
          >
            <Icon
              type="close"
            >
              <i
                className="kdicon kdicon-close"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </div>
        </div>
        <div
          className="kd-drawer-body"
        />
      </div>
    </div>
  </Portal>
</Drawer>
`;

exports[`Drawer render null or undefined without errors 1`] = `
<Drawer>
  <Portal
    containerInfo={
      <body>
        <div
          class="kd-drawer kd-drawer-hide"
          style="z-index: 1050;"
        >
          <div
            class="kd-drawer-mask"
          />
          <div
            class="kd-drawer-container-box kd-drawer-has-container-box kd-drawer-container-right"
            style="right: 0px; width: 320px; transform: translateX(100%);"
          >
            <div
              class="kd-drawer-header"
            >
              <div
                class="kd-drawer-title"
              />
              <div
                class="kd-drawer-close-icon"
              >
                <i
                  class="kdicon kdicon-close"
                />
              </div>
            </div>
            <div
              class="kd-drawer-body"
            />
          </div>
        </div>
        <div
          class="kd-drawer kd-drawer-hide"
          style="z-index: 1050;"
        >
          <div
            class="kd-drawer-mask"
          />
          <div
            class="kd-drawer-container-box kd-drawer-has-container-box kd-drawer-container-right"
            style="right: 0px; width: 320px; transform: translateX(100%);"
          >
            <div
              class="kd-drawer-header"
            >
              <div
                class="kd-drawer-title"
              />
              <div
                class="kd-drawer-close-icon"
              >
                <i
                  class="kdicon kdicon-close"
                />
              </div>
            </div>
            <div
              class="kd-drawer-body"
            />
          </div>
        </div>
      </body>
    }
  >
    <div
      className="kd-drawer kd-drawer-hide"
      style={
        Object {
          "zIndex": 1050,
        }
      }
    >
      <div
        className="kd-drawer-mask"
        onClick={[Function]}
      />
      <div
        className="kd-drawer-container-box kd-drawer-has-container-box kd-drawer-container-right"
        style={
          Object {
            "right": 0,
            "transform": "translateX(100%)",
            "width": 320,
          }
        }
      >
        <div
          className="kd-drawer-header"
        >
          <div
            className="kd-drawer-title"
          />
          <div
            className="kd-drawer-close-icon"
            onClick={[Function]}
          >
            <Icon
              type="close"
            >
              <i
                className="kdicon kdicon-close"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </div>
        </div>
        <div
          className="kd-drawer-body"
        />
      </div>
    </div>
  </Portal>
</Drawer>
`;

exports[`Drawer renders correctly 1`] = `
<div
  class="kd-drawer kd-drawer-hide kd-drawer-not-at-body"
  style="z-index:1050"
>
  <div
    class="kd-drawer-mask"
  />
  <div
    class="kd-drawer-container-box kd-drawer-container-right"
    style="right:0;width:320px;transform:translateX(100%)"
  >
    <div
      class="kd-drawer-header"
    >
      <div
        class="kd-drawer-title"
      />
      <div
        class="kd-drawer-close-icon"
      >
        <i
          class="kdicon kdicon-close"
        />
      </div>
    </div>
    <div
      class="kd-drawer-body"
    />
  </div>
</div>
`;

exports[`Drawer renders correctly 2`] = `
<div
  class="kd-drawer kd-drawer-hide kd-drawer-not-at-body"
  style="z-index:1050"
>
  <div
    class="kd-drawer-mask"
  />
  <div
    class="kd-drawer-container-box kd-drawer-container-top"
    style="top:0;transform:translateY(-100%)"
  >
    <div
      class="kd-drawer-header"
    >
      <div
        class="kd-drawer-title"
      />
      <div
        class="kd-drawer-close-icon"
      >
        <i
          class="kdicon kdicon-close"
        />
      </div>
    </div>
    <div
      class="kd-drawer-body"
    />
  </div>
</div>
`;

exports[`Drawer renders correctly 3`] = `
<div
  class="kd-drawer kd-drawer-hide kd-drawer-not-at-body"
  style="z-index:1050"
>
  <div
    class="kd-drawer-mask"
  />
  <div
    class="kd-drawer-container-box kd-drawer-container-right"
    style="right:0;width:320px;transform:translateX(100%)"
  >
    <div
      class="kd-drawer-header"
    >
      <div
        class="kd-drawer-title"
      />
      <div
        class="kd-drawer-close-icon"
      >
        <i
          class="kdicon kdicon-close"
        />
      </div>
    </div>
    <div
      class="kd-drawer-body"
    />
  </div>
</div>
`;

exports[`Drawer renders correctly 4`] = `
<div
  class="kd-drawer kd-drawer-hide kd-drawer-not-at-body"
  style="z-index:1050"
>
  <div
    class="kd-drawer-mask"
  />
  <div
    class="kd-drawer-container-box kd-drawer-container-bottom"
    style="bottom:0;transform:translateY(100%)"
  >
    <div
      class="kd-drawer-header"
    >
      <div
        class="kd-drawer-title"
      />
      <div
        class="kd-drawer-close-icon"
      >
        <i
          class="kdicon kdicon-close"
        />
      </div>
    </div>
    <div
      class="kd-drawer-body"
    />
  </div>
</div>
`;

exports[`Drawer renders correctly 5`] = `
<div
  class="kd-drawer kd-drawer-hide kd-drawer-not-at-body"
  style="z-index:1050"
>
  <div
    class="kd-drawer-mask"
  />
  <div
    class="kd-drawer-container-box kd-drawer-container-left"
    style="left:0;width:320px;transform:translateX(-100%)"
  >
    <div
      class="kd-drawer-header"
    >
      <div
        class="kd-drawer-title"
      />
      <div
        class="kd-drawer-close-icon"
      >
        <i
          class="kdicon kdicon-close"
        />
      </div>
    </div>
    <div
      class="kd-drawer-body"
    />
  </div>
</div>
`;
